src/sgl/device/shader_object.h

```diff
@@ -49,6 +51,7 @@ class SGL_API ShaderObject : public Object {
     virtual void get_cuda_interop_buffers(std::vector<ref<cuda::InteropBuffer>>& cuda_interop_buffers) const;
 
     rhi::IShaderObject* rhi_shader_object() const { return m_shader_object; }
+    ref<Device> get_device() const { return m_device; }
```

Comment:
in sgl we use just the property name for the getters in general .. so `device()` instead of `get_device()`

Reply:
also, you might just want to return `Device*` to not incur ref-counting overhead